bittide-instances/src/Bittide/Instances/Hitl/HwCcTopologies.hs

```diff
@@ -488,54 +632,141 @@ hwCcTopologyTest refClkDiff sysClkDiff syncIn rxns rxps miso =
+                         $ natToNum @(PeriodToCycles Basic125 AllStablePeriod)
+    }
+
+  calibrateClockOffsets =
+    ( -- the name must be chosen such that this test is executed first
+      "0_calibrate_clock_offsets"
+    , ( toList $ imap (,) $ repeat @FpgaCount TestConfig
+          { fpgaEnabled       = True
+          , calibrate         = True
+          , stepSizeSelect    = commonStepSizeSelect
+          , initialClockShift = 0
+          , startupDelay      = 0
+          , mask              = maxBound
+          }
+      , defSimCfg
+          { mTopologyType  = Just $ Complete $ natToInteger @FpgaCount
+          , clockOffsets   = toList $ repeat @FpgaCount 0
+          , startupOffsets = toList $ repeat @FpgaCount 0
+          }
+      )
+    )
+
+  -- tests the given topology
+  tt ::
     forall n.
     (KnownNat n, n <= FpgaCount) =>
-    Topology n -> (TestName, (Probes TestConfig, SimConf))
-  testTopology t =
+    Vec n InitialClockShift ->
+    Vec n StartupDelay ->
+    Topology n ->
+    (TestName, (Probes TestConfig, SimConf))
+  tt clockShifts startupDelays t =
     ( fromString $ topologyName t
-    , ( toList (imap testData $ linkMasks @n t)
-         <> [ (fromInteger i, disabled)
-            | i <- [natToNum @n, natToNum @n + 1 .. natToNum @(FpgaCount - 1)]
-            ]
-      , def { mTopologyType = Just $ topologyType t
-            , simulationSamples = 1000
-            , simulationSteps = natToNum @(PeriodToCycles Basic125 (Seconds 60))
-            , stabilityMargin = snatToNum cccStabilityCheckerMargin
-            , stabilityFrameSize = snatToNum cccStabilityCheckerFramesize
-            , disableReframing = not $ cccEnableReframing
-            , rusty = cccEnableRustySimulation
-            , waitTime = fromEnum cccReframingWaitTime
-            , clockOffsets = toList $ repeat @n 0
-            , startupOffsets = toList $ repeat @n 0
-            , stopAfterStable =
-                Just $ natToNum @(PeriodToCycles Basic125 AllStablePeriod)
-            }
+    , ( toList ( zipWith4 testData indicesI clockShifts startupDelays
+               $ linkMasks @n t
+               )
+          <> [ (fromInteger i, disabled)
+             | let n = natToNum @n
+             , i <- [n, n + 1 .. natToNum @(FpgaCount - 1)]
+             ]
+      , let
+          clkPeriod :: Num a => a
+          clkPeriod = case clockControlConfig of
+            (_ :: ClockControlConfig dom a b c) ->
+              snatToNum (clockPeriod @dom)
+          stepSizeDiv = case commonStepSizeSelect of
+            PPB_1   -> 1_000_000
+            PPB_10  ->   100_000
+            PPB_100 ->    10_000
+            PPM_1   ->     1_000
+          conv s d x = ((/ d) . (s *)) . fromIntegral <$> toList x
+        in
+          defSimCfg { mTopologyType  = Just $ topologyType t
+                    , clockOffsets   = conv clkPeriod stepSizeDiv clockShifts
+                    , startupOffsets = conv (1000 * clkPeriod) 1 startupDelays
```

Comment:
To avoid loss in precision via first multiplying by `1000` and then dividing again at least by a million due by ppm/ppb. `stepSizeDiv` already has been reduced by that order of magnitude accordingly.

Reply:
Ah... yeah that deserves an `-- XXX: ` comment. Working around floating point precision loss, always a joy.

Reply:
> `stepSizeDiv` already has been reduced by that order of magnitude accordingly.

I think this should also be documented. Its confusing to have 1 function that takes both femtoseconds and picoseconds, but works "by accident" because another function arguments happens to div it again.